### PR TITLE
11.0 http cache etag fix

### PR DESCRIPTION
**Pull request by ap-wtioit**
Description of the issue/feature this PR addresses:
* When updating modules static web xml templates (e.g. /website_twitter/static/src/xml/website.twitter.xml) the old content gets used for up to a week.

Current behavior before PR:
* XML Templates depended on by javascript Widgets get cached for a week in CDN and Browsers

Desired behavior after PR is merged:
* XML Templates depended on by javascript Widgets get reloaded soon after modules are updated in source.

HTTP/1.1 needs to be set for Chrome to use ETag Headers. New environment variable lets server admin decide which browser cache times to use. After browser cache timeout the browser will send a HTTP/1.1 request with If-None-Match header and odoo/werkzeug most of the times sends a quick 304 response. 

Info @wt-io-it

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by wtaferner** [comment]
Another example would be [Website Sale Stock Availabililty](https://github.com/odoo/odoo/blob/13.0/addons/website_sale_stock/static/src/xml/website_sale_stock_product_availability.xml) and [the related loading via JS](https://github.com/odoo/odoo/blob/13.0/addons/website_sale_stock/static/src/js/variant_mixin.js#L10)

Any of these resources loaded via the static folder and Javascript Async with some frontend rendering are exposed to this issue in terms of cache busting.

Still there in 13.0 and probably worth to provide a fix for Odoo core that the developer does not need to take care of cache busting. 

@odony @mart-e @JKE-be 
Not sure who wants to review and maybe in best case would merge this as it is or wants to elaborate to provide an even better solution for the issue we have here. Let us know how we can help to understand and reproduce to make this happen.

**Comment by odony** [comment]
I doubt we need a patch in the core for this. If your issue is with CDN caching, you should handle the cache policy at the CDN level.

And even without a CDN, you should have a web server in front of Odoo (required for SSL termination anyway), and you should be able to configure it to use any protocol version and return any appropriate headers for /*/static files if you like.

E.g. for nginx you could have this in your server config:

```
    location ~ ^/[^/]+/static/.+\.(js|xml)$ {
        expires 1h;
        add_header Access-Control-Allow-Origin '*';
        add_header X-Content-Type-Options nosniff;
        proxy_pass ... ;
    }
```

Basically we don't want to have a default cache value that is too short, as most deployments will be perfectly fine with a 1 week cache. 
So if it has to be configured on a case-by-case basis, why not do it at the appropriate place: the web server or CDN config?
We try to keep the HTTP boilerplate in Odoo to a generic minimum (e.g. sessions, etc.), we don't want to reimplement a web server, and werkzeug is clearly not designed for that anyway ;-)

**Comment by ap-wtioit** [comment]
@odony that requires that you have nginx before odoo, the issue is in odoo itself. it sends wrong headers.

is it a official requirement to have a proxy that fixes the odoo headers in front of odoo?

The first issue is:
Odoo sends ETag Headers but also responds with HTTP/1.0 (which is invalid and chrome ignores ETag Headers when the server responds with HTTP/1.0, you need HTTP/1.1 for ETag to work in chrome)

The second issue is:
Odoo defaults to send a week as browser cache timeout with no chance to configure it. If you want lower values or higher values you need a proxy in front of it.

Our issue is not with CDN caching solely but with all caching in front of odoo (CDN, proxy, browser) because odoo sends headers that are wrong and not configurable.

**Comment by odony** [comment]
> is it a official requirement to have a proxy that fixes the odoo headers in front of odoo?

Yes it is, how else do you do SSL?

> Our issue is not with CDN caching solely but with all caching in front of odoo (CDN, proxy, browser) because odoo sends headers that are wrong and not configurable.

Odoo sends default headers and is not meant to be configurable to the level of a web server, I just explained why.
